utils/clients/coingecko/client.go

```diff
@@ -289,3 +290,114 @@ func (c *HTTPClient) FetchSupportedVsCurrencies(ctx context.Context) (*Supported
 
 	return &body, nil
 }
+
+type coinMarketParams struct {
+	baseParams
+	VsCurrency string `url:"vs_currency"`
+	Page       int    `url:"page"`
+	PerPage    int    `url:"per_page"`
+	Limit      int    `url:"-"`
+}
+
+// GenerateQueryString - implement the QueryStringBody interface
+func (p *coinMarketParams) GenerateQueryString() (url.Values, error) {
+	p.Page = 1
+	p.PerPage = int(p.Limit)
+	return query.Values(p)
+}
+
+// CoinMarket represents the market data for a single coin returned in
+// FetchCoinMarkets call to coingecko
+type CoinMarket struct {
+	ID                       string  `json:"id"`
+	Symbol                   string  `json:"symbol"`
+	Name                     string  `json:"name"`
+	Image                    string  `json:"image"`
+	MarketCap                int     `json:"market_cap"`
+	MarketCapRank            int     `json:"market_cap_rank"`
+	CurrentPrice             int     `json:"current_price"`
+	PriceChange24h           float64 `json:"price_change_24h"`
+	PriceChangePercentage24h float64 `json:"price_change_percentage_24h"`
+	TotalVolume              int     `json:"total_volume"`
+}
+
+// CoinMarketResponse is the coingecko response for FetchCoinMarkets
+type CoinMarketResponse []CoinMarket
+
+// FetchCoinMarkets fetches the market data for the top coins
+func (c *HTTPClient) FetchCoinMarkets(
+	ctx context.Context,
+	vsCurrency string,
+	limit int,
+) (*CoinMarketResponse, time.Time, error) {
+	updated := time.Now()
+	url := "/api/v3/coins/markets"
+	params := &coinMarketParams{
+		baseParams: c.baseParams,
+		VsCurrency: vsCurrency,
+		Limit:      limit,
+	}
+
+	cacheKey, err := c.cacheKey(ctx, url, params)
```

Comment:
Would it be better to just always get from CG the top 250 and cache that?  As I read this, we are using the limit in the cache key.

Suggest always getting top 250 from CG and then in only return the `limit` number of responses from what has been cached.

Reply:
Yes, that's better, will adjust the logic.  Thanks

Reply:
2f2d1e38f0d9b050683711f1e2b0e94b62011d55